Preload extension: hx-boost, no side effects, HX-Preloaded header, form support, tests and JSDocs

marisst opened the pull request:
Major update of the [preload extension](https://github.com/marisst/htmx-extensions/blob/main/src/preload/preload.js).

- **Hx-boost element preloading with correct request headers**
    - Currently `<a href="link" hx-boost="true" preload>Link</a>` will be preloaded using XML request without `HX-Request` headers.
    - Hoverer, when user clicks on the link, a request with `HX-Request` and `HX-Boosted` headers is sent, and the cached request may not be used by the browser resulting in double loading.
    - This update includes changes fixing this issue and also allowing to preload `<form method="get" target="/link" hx-boost="true" preload>` with `HX-Request`.
    - Issues addressed:
        - #89 
        - #107 
        - #108
        - #116
- **No side-effects e.g. like visible loading indicators while preloading data**
    - Currently if an element is preloaded based on `hx-get` attribute, the preload will cause loading indicators to show up
    - With this change it is not the case anymore. `HTMX.ajax()` is used only to generate the `xhr` request (incl. headers), which is manipulated in `htmx:beforeRequest` and sent without executing HTMX callbacks.
    - Issues addressed:
        - #78
        - #90
        - #110 
        - #115
    - Overlapping PR:
        - https://github.com/bigskysoftware/htmx-extensions/pull/91
- **HX-Preload header**
    - Currently server cannot distinct between preload and regular requests
    - With this change, a `HX-Preload` header is added to all preload requests
    - Issues addressed:
       - https://github.com/bigskysoftware/htmx-extensions/issues/114
    - Overlapping PR:
        - https://github.com/bigskysoftware/htmx-extensions/pull/61
- **Form element smart preloading with alternated values**
    - Currently `<form hx-get="/link" preload>` will preload the form with as-is values whenever user clicks or hovers (if `preload="mouseover"`) on any of the forms elements.
    - However, if user clicks or hovers over `<input type="radio">`, `<input type="checkbox">` or `<select>` elements, it is more likely that the value of those elements will be changed first and then the form submitted.
    - This update includes smarter preloading of these three input elements, sending preload request with the alternated values. Additionally, `<form method="get" target="/link" preload>` will also be preloaded. Elements other than radio buttons, checkboxes, select elements and submit buttons are not triggering form preloading.
    - Issue addressed: bigskysoftware/htmx#105 
- **Tests**
    - Currently extension doesn't have any tests
    - Added [44 tests](https://github.com/marisst/htmx-extensions/tree/main/src/preload/test/ext) covering current and added functionality
    - Two issues were found and fixed during this process
        - `<button hx-get="/link" hx-ext="preload" preload>Button</button>` is not preloaded in current version, but it is preloaded with the change. The reason is that the extension currently does not consider the element in which the extension is registered among the pool of nodes it initializes.
        - `touchstart` event listener was added to node as an alternative to `mousedown` and `mouseover` events for touchscreens, but it was not working because the correct `node.preloadState` was not set when calling `load(node)`
- **JSDocs and improved code readability**
    - JSDocs added to all functions
    - Code reorganized in top-down manner. Currently the main part of the code is on the bottom and the least important utility functions are on top, making it hard to read it. I replaced `const funct = function(){}` with `function funct() {}` declarations, which allow to change the order.
    - The `node.preloadState` values are currently confusing. `PAUSE` is used for when node is initialized, `READY` for when data is loading and `DONE` for when loading is complete. I have renamed `PAUSE` to `READY`, `READY` to `LOADING` and kept `DONE`. I also added a new state `TIMEOUT` used to check that user hovers over an element for at least 100ms before it starts loading.
- **`preload="always"` documentation and config flexibility**
    - added documentation for `preload="always"` property
    - enabled use of `preload="always"`, currently it can only be used together with preload mode e.g. `preload="always mousedown"` 

Documentation is updated in a corresponding pull request in the main xtmx repository. https://github.com/bigskysoftware/htmx/pull/3001

untitaker commented:
this patch is working great for me in combination with hx-boost. essentially it "just works" while the main branch version doesn't. however, the loading indicator is visible when pages are being preloaded.

ALameLlama commented:
Looks greats so far but +1 on the indicators, I don't believe these should be visible for preloads.

marisst commented:
@untitaker and @ALameLlama, thank you for reviewing my pull request. I agree with you. I implemented a change which prevents loading indicators to be shown while preloading nodes. I have also added a `"HX-Preload": "true"` header to all preload requests.

MichaelWest22 commented on src/preload/preload.js:
```diff
@@ -1,141 +1,386 @@
+  * The extension relies on browser cache and for it to work
+  * server response must include `Cache-Control` header
+  * e.g. `Cache-Control: private, max-age=60`.
+  * For more details @see https://v2-0v2-0.htmx.org/extensions/preload/
```
https://v2-0v2-0.htmx.org/extensions/preload/ should be https://htmx.org/extensions/preload/

The v2 sub site was just a snapshot of the main htmx.org site that was used during v2 development and early publication but the main site now has all the current v2 documentation.

marisst replied:
Thank you @MichaelWest22, I have updated the link.

ALameLlama commented on src/preload/preload.js:
```diff
@@ -1,141 +1,386 @@
+    // such as showing loading indicators while preloading data. 
+    if (name === 'htmx:beforeRequest') {
+      const requestHeaders = event.detail.requestConfig.headers
+      if (!("HX-Preload" in requestHeaders
+            && requestHeaders["HX-Preload"] === "true")) {
```
For continuity, I believe this should be HX-Preloaded instead of HX-Preload, similar to how adding hx-boost will add the header HX-Boosted
https://htmx.org/reference/#request_headers

marisst replied:
@ALameLlama Nicholas I agree. I have renamed it.

yoonthegoon commented:
I hate to be the bearer of bad news, but I'm still getting two requests and the latter not being cached.  
I attached screenshots on two browsers I've tested on.  
I'm on `macOS 15.0 24A335 arm64`.

I suspect it's due to the additional `HX-Preloaded` header, as that's the only difference between the two requests.  

```diff
GET /blog/inquit-incepta HTTP/1.1
Host: 127.0.0.1:3000
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:132.0) Gecko/20100101 Firefox/132.0
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br, zstd
HX-Request: true
HX-Current-URL: http://127.0.0.1:3000/blog
HX-Boosted: true
- HX-Preloaded: true
Sec-GPC: 1
Connection: keep-alive
Referer: http://127.0.0.1:3000/blog
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-origin
DNT: 1
Priority: u=0
```

Unfortunately, I wasn't able to modify your code to try removing the added header without breaking the way the script functioned.  
I'll see if I can try again on a different operating system when I get the chance.

I left both screenshots filtered just to show other requests were being cached.

Zen Browser  
`1.0.1-a.19 (Firefox 132.0.1) (aarch64)`
<img width="1081" alt="image" src="https://github.com/user-attachments/assets/b86cd66c-c99b-46c7-982f-379163113503">

Safari  
`Version 18.0 (20619.1.26.31.6)`
<img width="685" alt="image" src="https://github.com/user-attachments/assets/905b6198-1caa-4857-a117-e1fbfcdeac48">

ALameLlama commented:
What do your response headers look like? I have a feeling you're missing the Cache-Control.

https://htmx.org/extensions/preload/
> Preloaded responses will only be cached in the browser if the response headers allow it. For example, the response header Cache-Control: private, max-age=60 allows the browser to cache the response, whereas Cache-Control: no-cache prevents it.

The new request header HX-Preloaded shouldn't have any effect on it. just give the BE more information about what triggered the request.

yoonthegoon commented:
Oof I'm silly. I just added the Cache-Control header and it's all good now. (I blame axum and askama_axum for making adding that in so difficult :P )

Here was my old response headers.

```diff
HTTP/1.1 200 OK
content-type: text/html; charset=utf-8
content-length: 5740
- date: Thu, 21 Nov 2024 04:22:51 GMT
+ date: Thu, 21 Nov 2024 04:22:52 GMT
```

Also I tried the `preload.js` in https://github.com/bigskysoftware/htmx-extensions/pull/106/commits/6944833c62de97138a48b8dbda9ee05e73dfa058 (where the request headers are exactly the same) and I still wasn't caching the second, so that clearly wasn't it.

marisst commented:
It would be great to get somebody review this pull request. Merging it would allow us to close 10 issues, which is a lot given there are only roughly 30 issues reported in this repository.

@Telroshan seems that the most recent PRs in the repo have been reviewed by you. Have you had a chance or are you planning to look at this one? Do you have any worries about it? If helpful, we can connect.

The author of the original preload extension is @benpate https://github.com/bigskysoftware/htmx/pull/273, and it was approved by @1cg. Maybe we can get some inputs from the original authors if they are still active in this repo?

Telroshan commented:
Hey @marisst , first of all thanks for your work. Life's currently a bit hectic so I haven't had the chance yet to sit down and review this PR as it should be. Can't guarantee a date that I'll look into it, but I hope to do it in the upcoming weeks.
I know this can be frustrating, but rest assured that we'll get there eventually!

Telroshan reviewed (commented):
Hey, thanks a lot for your work!
Just as I was reading the previous extension's code before looking at this PR, I thought "how could this `touchstart` trick work given that it doesn't properly set the state?", and I was super glad to find out you had already identified that very issue in the PR's description, and fixed it 😁 

- I've left a few comments, many are simply nitpicking, or things I'm unsure about, so feel free to tell me to back off on those 😄 
- When we are over with all those comments, I'll try to get 1cg to either review this PR too, or trust my sole judgment, just be aware that we _might_ ask you to cut down your code changes to simplify the diff & reduce the cognitive load for him to review this PR.
I'm thinking that for example, the "smart preload" of form inputs could be done in a separate PR to alleviate this one and get it reviewed & merged faster

(Note: I haven't looked at the tests yet, I'm stopping here for today)

marisst commented on src/preload/package.json:
```diff
@@ -1,7 +1,8 @@
   "name": "htmx-ext-preload",
   "main": "preload.js",
-  "version": "2.0.1",
+  "version": "2.0.3",
```
I have updated the version number to `2.0.1`.

marisst commented on src/preload/package.json:
```diff
@@ -1,7 +1,8 @@
   "main": "preload.js",
-  "version": "2.0.1",
+  "version": "2.0.3",
+  "type": "module",
```
I have removed `"type": "module"` from the package configuration file.

marisst commented on src/preload/test/.DS_Store:
I have removed this and the other `.DS_Store` file which I had accidentally commited. I have added `.DS_Store` to `.gitignore` file to avoid this happening in the future. `.DS_Store` is a file where Mac stores custom information about how a folder in the file system should be displayed e.g. folder view options and icon positions. This has nothing to do with any of the extensions and therefore belongs in `.gitignore`,

marisst commented on src/preload/preload.js:
I have followed your suggestion and enclosed everything in anonymous function, following how it is done in the other extensions.

marisst commented on src/preload/preload.js:
```diff
@@ -1,141 +1,386 @@
+  const isPreloadableGetFormElement = node.form
+                                      && includesGetRequest(node.form)
+                                      && isPreloadableFromElement(node)
+  if (!includesGetRequest(node) && !isPreloadableFromElement) {
```
Thank you - you are completely right! I have fixed it.

marisst commented on src/preload/preload.js:
```diff
@@ -1,141 +1,386 @@
+  // or preloadable "GET" form elements
+  const getReqAttrs = ['href', 'hx-get', 'data-hx-get'];
+  const includesGetRequest = node => getReqAttrs.some(a => node.hasAttribute(a))
+                                      || node.method == 'get';
```
Done and done. :)

marisst commented on src/preload/preload.js:
```diff
@@ -1,141 +1,386 @@
+  if (node instanceof HTMLInputElement) {
+    const preloadableInputTypes = ['checkbox', 'radio', 'submit'];
+    return preloadableInputTypes.some(
+      type => node.getAttribute('type') === type);
```
You are right - this code is too complex. I have simplified it to make sure `getAttribute` is called only once. I have replaced `some` with `includes`. I believe `includes` is more elegant than `indexOf`, which you proposed, because we don't actually use the index anywhere. We just want to check if the type is among the types for which the extension currently supports smart preloading.

Telroshan replied:
Yep you're totally right, as we were coding in a IE-11 compatible way for htmx 1 (which doesn't support include), I still have some IE-11 friendly code habits sometimes, sigh... `includes` is perfect here!

marisst commented on src/preload/preload.js:
```diff
@@ -1,141 +1,386 @@
+    return node.control && isPreloadableFromElement(node.control);
+  }
+  return node instanceof HTMLSelectElement
+    || node.type === 'submit';
```
You are right that the `input type="submit"` is already handled in the first `if` of the function. Therefore the `node.type === 'submit'` statement in the highlighted line is redundant, and I have therefore removed it.

I don't think that we should preload other `button` elements than submit buttons. They don't necessary imply a server request and might be used e.g. to clear all form fields.

Telroshan replied:
Likely poorly formulated on my end, sorry! I indeed meant combining a `button` instance check + a `type` attribute check (not solely the `button` check)

marisst commented on src/preload/preload.js:
```diff
@@ -1,141 +1,386 @@
-          break
+    // radio and checkbox
+    const inputType = node.getAttribute("Type") || node.control.getAttribute("Type");
+    const nodeValue = node.value || node.control.value;
```
Thank you - I solved this with optional chaining `node.control?.value`.

marisst commented:
@Telroshan thank you for dedicating time and reviewing my changes diligently! ☺️ 

I have fixed all the issues you found in the way you proposed, except for the following four issues which I solved in a different way. I have left more detailed explanations as replies to the relevant comments.
- Did not add smart form element preloading to all `button` elements, as suggested in one of the comments. Only to `submit` and `button type="submit"`. Other buttons might not imply a server request and it is unclear how we would preload them e.g. button to clear form fields.
- Replaced the use of `some` with `includes`. You suggested using `indexOf`. I believe `includes` is a better fit because we are not using the index anywhere, and only need a boolean check.
- Instead of proposed `node.control && node.control.value` I do optional chaining `node.control?.value` with `?`.
- Added `.DS_STORE` to `.gitignore` to prevent accidentally adding Mac OS specific files to the repository in the future.

Telroshan commented on src/preload/preload.js:
```diff
@@ -1,141 +1,387 @@
+
+    // Don't preload <input> elements contained in <label>
+    // to prevent sending two requests
+    if (node.tagName instanceof HTMLInputElement && node.closest('label')) {
```
Seems you forgot to remove the `tagName` access here, as the `instanceof` check is currently performed on a string rather than the node itself
(Also, the fact that this mistake doesn't fail any test in the current test suite, _might_ - unsure as I haven't reviewed the tests yet - reveal that a test case is missing to check this specific case?)

marisst replied:
@Telroshan Yes, I was too fast here. 😆  Fixed and added test for it in https://github.com/bigskysoftware/htmx-extensions/pull/106/commits/72c2fe25741ff8a16f084ca328bd97ac4cb58826.

Telroshan commented on src/preload/preload.js:
```diff
@@ -1,141 +1,387 @@
+    if (node instanceof HTMLLabelElement) {
+      return node.control && isPreloadableFormElement(node.control);
+    }
+    return node instanceof HTMLSelectElement;
```
I likely poorly formulated what I meant ; since we have a `instanceof HTMLInputElement` check above, and the previous code was checking `node.type === 'submit'` here, I meant that we could add a button `instanceof` check to make the code clearer on first read (so just for clarity, doesn't change the behavior).

The issue now that it has been removed, is that a `button type="submit"` would, unless I'm mistaken, not be handled anymore, as it wouldn't fall through the `instanceof HTMLInputElement` check

I guess you could either replace `if (node instanceof HTMLInputElement) {` by `if (node instanceof HTMLInputElement || node instanceof HTMLButtonElement ) {`,
or, replace
`return node instanceof HTMLSelectElement;` by `return node instanceof HTMLSelectElement || (node instanceof HTMLButtonElement && node.type === 'submit');`

If I'm not wrong and `submit` buttons were indeed not handled anymore due to the latest removal, this _might_ - unsure as I haven't reviewed the tests yet - reveal that a test case is missing to check this specific case, as it currently doesn't fail any test in the test suite,

Telroshan commented on src/preload/test/ext/repeated-interaction.js:
I'm thinking you might want to add a `mouseover always` test too, just to be extra very 100% sure that it works with other events than `mousedown` + ensure it works properly with the 100 ms delay

marisst replied:
Thank you. I added a test for `mouseover always` repeated interaction in https://github.com/bigskysoftware/htmx-extensions/pull/106/commits/fd14c11be9db6d568cfec32753e368e14b3ef364. Additionally I also added a test for `always mousedown` to make sure the order of the keywords doesn't break anything and it behaves the same way as `mousedown always`.

Telroshan commented on src/preload/test/ext/requests.js:
```diff
@@ -0,0 +1,148 @@
+      should.equal(requests.length, 2)
+      should.equal(requests[0].url, requests[1].url)
+      should.equal(requests[0].method, requests[1].method)
+      requests[0].requestHeaders.should.deep.contain(requests[1].requestHeaders)
```
Maybe we should also check, on top of the preloaded request containing all the headers of the htmx click request, that the click request does **not** include the preloaded header? Just to ensure we don't mess this up in the future if we were to make new changes to this extension over time

marisst replied:
Here you go https://github.com/bigskysoftware/htmx-extensions/pull/106/commits/2f105eff533e2e2099c94d7249113f0e5557fdb2 😉

Telroshan reviewed (commented):
Excellent test suite!

Telroshan reviewed (approved):
Alright, looks perfect to me!
I'll discuss with 1cg and keep you informed

Telroshan commented:
Got 1cg's approval, here it goes! Thanks again for your amazing work on this!
